libs/parallelism/execution/include/hpx/execution/sender_future.hpp

```diff
@@ -0,0 +1,137 @@
+//  Copyright (c) 2021 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/allocator_support/allocator_deleter.hpp>
+#include <hpx/allocator_support/internal_allocator.hpp>
+#include <hpx/execution/algorithms/detail/single_result.hpp>
+#include <hpx/execution_base/operation_state.hpp>
+#include <hpx/execution_base/sender.hpp>
+#include <hpx/functional/invoke_result.hpp>
+#include <hpx/futures/detail/future_data.hpp>
+#include <hpx/futures/promise.hpp>
+#include <hpx/memory/intrusive_ptr.hpp>
+#include <hpx/type_support/pack.hpp>
+#include <hpx/type_support/unused.hpp>
+
+namespace hpx { namespace execution { namespace experimental {
+    namespace detail {
+        template <typename T, typename Allocator, typename OS>
+        struct future_data
+          : hpx::lcos::detail::future_data_allocator<T, Allocator>
+        {
+            HPX_NON_COPYABLE(future_data);
+
+            using operation_state_type = std::decay_t<OS>;
+            using init_no_addref =
+                typename hpx::lcos::detail::future_data_allocator<T,
+                    Allocator>::init_no_addref;
+            using other_allocator = typename std::allocator_traits<
+                Allocator>::template rebind_alloc<future_data>;
+
+            operation_state_type os;
+
+            template <typename S, typename R>
+            future_data(init_no_addref no_addref, other_allocator const& alloc,
+                S&& s, R&& r)
+              : hpx::lcos::detail::future_data_allocator<T, Allocator>(
+                    no_addref, alloc)
+              , os(hpx::execution::experimental::connect(
+                    std::forward<S>(s), std::forward<R>(r)))
+            {
+                hpx::lcos::detail::intrusive_ptr_add_ref(this);
+                hpx::execution::experimental::start(os);
+            }
+        };
+
+        template <typename T, typename Allocator>
+        struct make_future_receiver
+        {
+            hpx::lcos::detail::future_data_allocator<T, Allocator>* data;
```

Comment:
you could use a `hpx::intrusive_ptr<future_data_refcnt_base>` here instead (and a `data.release()` instead of the `intrusive_ptr_release` below. This would make things more robust wrt exception handling. Also, the `make_future_receiver` wouldn't have to be a template in this case.

Reply:
That sounds like a nice simplification, thanks!

Reply:
I pushed some more changes. It can't actually be only an intrusive pointer of `future_data_refcnt_base` since it needs to set the value, but an intrusive pointer of `future_data_allocator` still improves the exception handling.

Reply:
This looks good now, thanks for your work on this!